mattnenterprise/rust-ftp
# Upgrade to latest OpenSSL

This update is a bit messy, I appologize in advanced for the lack of structure.  I ended up doing a lot of hacking to try to address a few bugs I was dealing with in my own code.

Major things I've done here include:

* Updating to the latest Rust-OpenSSL
    * This fixes a bug wherein creating secure data connections would fail, due to a segfault in rust-openssl (or openssl itself)
* Switch the type contained by `FtpError::SecureError` from `SslError` to `Box<Error>` since the latest rust-OpenSSL contains several different errors in different modules, but we don't want to have to distinguish them
* Have `FtpStream` now contain an `SslContext` since it implements `Clone`, and `Ssl` doesn't
* Add an FTP server written in Python (using pyftpdlib) to test rust-ftp against
* Add a certificate and key to test TLS (FTPS) against
* Add an example program that uses the test cert to connect to the FTP server included to make sure encrypted control and data channels work

I know it's not great to have things all mixed up like this and again, I'm sorry for the lack of organization.  Since I use this library for work, a lot of what I'm doing with it is being added in incrementally according to my immediate needs.  I hope you'll agree that, even with an FTP server written in Python instead of Rust, all of these additions add value to the project.

Let me know if there's anything I can do to make this better.
